### DISCUSSION TITLE
terminate?

### DISCUSSION THREAD

**Discussion by jcoolsen** [johnny-quesada-developer/easy-web-worker#1] [Ideas]
This is really a small but great project. Made things much easier.
One thing I miss: Either direct access to the underlying Worker or the ability to call terminate() on a worker. For the use case when a worker has an infinite loop or is running longer than we want it to for any other reason.

**Comment by johnny-quesada-developer** [comment]
Hi [jcoolsen](https://github.com/jcoolsen),

Thanks for trying out my package, I'm glad to hear that it's been helpful for you! I appreciate your feedback about the terminate functionality and agree that it's an important feature. I'll be sure to include it in the next commit.

Additionally, I'm planning on changing the encapsulation of the underlying worker to provide more direct access, which should further improve the package's usability.

Thanks again for your input and please feel free to let me know if you have any further suggestions or questions.

Best regards,

**Reply by johnny-quesada-developer** [reply]
Hi @jcoolsen ,

I wanted to let you know that your feedback about adding the terminate functionality to Easy Web Worker was implemented in the latest update. In addition, I've also made changes to the encapsulation of the underlying worker to provide more direct access.

I also wanted to mention that I've created an example repository on GitHub (https://github.com/johnny-quesada-developer/easy-web-workers-example) to help users get started with Easy Web Workers

Lastly, I've updated the messages to return a **CancelablePromise** from **cancelable-promise-jq**, which provides even more capabilities to the library.

Thank you again for your feedback, and please let me know if you have any further suggestions or questions.

Best regards,

**Reply by jcoolsen** [reply]
Thank you, @johnny-quesada-developer 
I appreciate this. Unfortunately having tested cancellation of promise and termination of worker (via the deprecated property), neither allow me to use the worker again after it has "run amok".
I am actually running jsdiff on two files with thousands of differences and the time complexity is high. I don't think there is anything wrong with the changes you've made in this project. It just seems that even terminating a worker cannot allow it to be reused in this case. It must be a general/browser/implementation thing.
One (other) thing I have been struggling with, is how to reference the jsdiff script. Since this feature is part of a spfx web part, everything is wrapped in obscure webpack and typescript inclusion "magic". I ended up referencing the the jsdiff from a public cdn, which brings me to my point: I assume that you load the referenced scripts into a blob and reference a bloburl from the worker, or maybe you embed them. Either way, and now I'm pushing it, I know: Would it be possible to provide a fallback script url in case the first one is not available? If you feel this is beyond the scope of the project, I understand completely. No worries. :-)

**Reply by johnny-quesada-developer** [reply]
Hi @jcoolsen ,

Thank you for your message. I understand, you've been experiencing some issues with how to reference the **jsdiff** library within your easy workers.

The worker being useless after the import fails provable is due the fact that the import is the first thing that happens, so if the imports fails the rest of the code is not gonna be initialized.

However, there are two possible solutions:

1-  Inside the template of the **easy-worker** you can use a try-catch block around the **self.importScripts** call to catch any errors and load an alternative script instead. Here's an example:

```ts
const worker = new EasyWebWorker<Payload>(
  (easy, context: DedicatedWorkerGlobalScope) => {
    try {
      context.importScripts('cdn/url');
    } catch (e) {
      context.importScripts('cdn/url');
      // Call alternative script or handle the error in another way
    }

    easy.onMessage((message) => {
      // do something...
    });
  }
);
```

2 - The second option required a little extra of boilerplate but maybe it will be the best solution for you since will help you to don't depend on the CDN.

 - in the webpack.config:
 ```ts
  entry: {
   ..,// other entries
    checkDiff: './src/checkDiff.worker.ts', // this is your new bundle
  },
   output: {
    ...,// other configs
    filename: ({ chunk: { name } }) => {
      if (name.includes('worker')) {
        return `${name}.worker.js`;
      }

      return `${name}.js`;
    },
  },
  ....
 ```
 
- then your static worker will be like this:

**checkDiff.worker.ts**
```ts
import something from 'jsdiff';
import { StaticEasyWebWorker } from 'StaticEasyWebWorker';

new StaticEasyWebWorker<unknown, unknown>(async (message) => {
  const { payload } = message;

  const result = await something(payload);

  message.resolve(result);
});

```
 - Finally, in the main thread, you can create a new instance of EasyWebWorker and point it to your worker script, like this:
 
 ```ts
 const worker = new EasyWebWorker('checkDiff.worker.js');
 ```
 
 All the capabilities of the easy web worker will work the same thanks to the **StaticEasyWebWorker**...

I hope this information helps. Please let me know if you have any further questions or concerns.

**Reply by jcoolsen** [reply]
Sorry for the delay @johnny-quesada-developer . Thank you for the suggestions, Option 1 is probably the best for me, since the webpack.config is hard to get at in my setup. I think that will work just great. Thank you for the effort, you're really going above a beyond.

**Reply by johnny-quesada-developer** [reply]
> Sorry for the delay @johnny-quesada-developer . Thank you for the suggestions, Option 1 is probably the best for me, since the webpack.config is hard to get at in my setup. I think that will work just great. Thank you for the effort, you're really going above a beyond.

No worries, thanks to you for the feedback 👍

If you are interested I created a live example with a similar library for comparing text, you could find the same here https://johnny-quesada-developer.github.io/easy-web-workers-example/ maybe could be of utility…

The only think is that I created the same using vite not webpack. 

Have a nice day 👍

**Comment by johnny-quesada-developer** [comment]
Hi @jcoolsen ,

The feedback was addressed. please the a look in the documentation [easy web worker methods](https://github.com/johnny-quesada-developer/easy-web-worker#easywebworkerrebootreason-unknown-cancelablecancelablepromisevoid)
